docs: update mocking docs with new override_during attribute (#36312)
hashicorp/terraform: website/docs/language/tests/mocking.mdx

An example of this is the `bucket` attribute in the `aws_s3_bucket` resource. A real AWS provider will generate a bucket name if one is not specified. A mocked AWS provider will do the same, and only generate a value if one is not already specified in the configuration.

> By default, Terraform will generate data during the `apply` operation while returning `(known after apply)` values during the `plan` operation. This behavior can be overridden with the `override_during` attribute in the `mock_provider` block:

Reviewer: ```suggestion
By default, Terraform generates data during the `apply` operation and returns `(known after apply)` values during the `plan` operation. You can override this behavior with the `override_during` attribute in the `mock_provider` block:
```

```hcl
mock_provider "aws" {
  override_during = plan
}
```

> The above `aws` provider will generate the data during the `plan` operation and reuse the same data during the `apply` operation. The `override_during` attribute accepts either `plan` or `apply` as values.

Reviewer: ```suggestion
The above `aws` provider generates the data during the `plan` operation and reuses the same data during the `apply` operation. The `override_during` attribute accepts either `plan` or `apply` as values.
```

### Mock Provider data

> You can specify specific values for targeted resources and data sources. in a `mock_provider` block, you can write any number of `mock_resource` and `mock_data` blocks. Both the `mock_resource` and `mock_data` blocks accept a type argument that should match the resource or data source you want to provide values for. They also accept a `defaults` object attribute that you can use to specify the values that should be returned for specific attributes.

Reviewer: ```suggestion
You can specify specific values for targeted resources and data sources. In a `mock_provider` block, you can write any number of `mock_resource` and `mock_data` blocks. Both the `mock_resource` and `mock_data` blocks accept a type argument that should match the resource or data source you want to provide values for. They also accept a `defaults` object attribute that you can use to specify the values that should be returned for specific attributes.
```

}
```

> In the above example, Terraform uses the supplied value for `arn` attributes in S3 buckets instead of generating a random string. Computed attributes not provided an explicit default will simply fall back to the generic data generation rules. The `override_during` attribute can also be used in the `mock_resource` and `mock_data` blocks to specify when the values should be generated at an individual resource level. If not specified, the values will be generated using the rules inherited from the `mock_provider` block. If specified, the local value will override any value specified in the `mock_provider` block.

Reviewer: ```suggestion
In the above example, Terraform uses the supplied value for `arn` attributes in S3 buckets instead of generating a random string. Computed attributes not provided an explicit default will simply fall back to the generic data generation rules. You can also use the `override_during` attribute in the `mock_resource` and `mock_data` blocks to specify when Terraform should generate the values for an individual resource. If you do not not specify the `override_during` attribute, Terraform generates the values using the rules inherited from the `mock_provider` block. If specified, the local value overrides any value specified in the `mock_provider` block.
```

Reviewer: Do we want to include an example of the `override_during` attribute?

Author: We do provide an example of using it in the `mock_provider` block, and the behaviour is the same across the three blocks so I didn't think it was needed - but I'm happy to be corrected.

Reviewer: Oh you're totally right! Sorry, I missed that. I think this is good as-is then.

### Overrides Syntax

> All override blocks contain a `target` attribute, which should specify the resource, data source, or module to override. The `override_module` blocks contain an `outputs` attribute, while the `override_resource` and `override_data` blocks contain a `values` attribute. In addition, the override blocks also support the same `override_during` attribute seen within the `mock_provider` block.

Reviewer: ```suggestion
All override blocks contain a `target` attribute, which should specify the resource, data source, or module to override. The `override_module` blocks contain an `outputs` attribute, while the `override_resource` and `override_data` blocks contain a `values` attribute. The override blocks also support the `override_during` attribute.
```

> The `outputs` and `values` attributes are optional and if not specified, Terraform will generate values for them automatically. The `override_during` attribute is also optional, and if specified it will override the `override_during` attribute in the `mock_provider` block. If not specified, the behavior will be inherited from the `mock_provider` block.

Reviewer: ```suggestion
The `outputs` and `values` attributes are optional and if not specified, Terraform will generate values for them automatically. The `override_during` attribute is also optional, and if specified it will override the `override_during` attribute in the `mock_provider` block. If not specified, Terraform inherits the behavior from the `mock_provider` block.
```
